llvm/lib/Transforms/Utils/PromoteMemoryToRegister.cpp

```diff
@@ -580,6 +580,10 @@ rewriteSingleStoreAlloca(AllocaInst *AI, AllocaInfo &Info, LargeBlockInfo &LBI,
       if (DbgItem->isAddressOfVariable()) {
         ConvertDebugDeclareToDebugValue(DbgItem, Info.OnlyStore, DIB);
         DbgItem->eraseFromParent();
+      } else if (DbgItem->isValueOfVariable() &&
+                 DbgItem->getExpression()->startsWithDeref()) {
+        InsertDebugValueAtStoreLoc(DbgItem, Info.OnlyStore, DIB);
```

Comment:
Doesn't this take the expression of the existing DbgItem, i.e. it will still have the deref even after mem2reg?

Reply:
Yes, but we are moving a `#dbg_val` so we want the deref

Reply:
But we're changing the debug value to point from the alloca to the value stored in the alloca right? In the same way that we replace a `dbg.declare` of the alloca with a `dbg.value` of the stored value, we should be replacing a `dbg.value` that derefs the alloca with a `dbg.value` of the stored value, without the deref.

Reply:
Hi @SLTozer you are correct, sorry about the late update, I got caught up in other stuff and forgot about this patch, the code has been changed to remove the deref from the expression in `InsertDebugValueAtStoreLoc`